src/cscapi/client.py

```diff
@@ -0,0 +1,276 @@
+import logging
+from typing import Dict, List, Tuple, Iterable
+from importlib import metadata
+
+import httpx
+import jwt
+from more_itertools import batched
+
+from cscapi.storage import MachineModel, ReceivedDecision, SignalModel, StorageInterface
+from dataclasses import replace
+
+__version__ = metadata.version("cscapi").split("+")[0]
+
+logging.getLogger("capi-py-sdk").addHandler(logging.NullHandler())
+
+CAPI_BASE_URL = "https://api.crowdsec.net/v3"
+CAPI_WATCHER_REGISTER_URL = f"{CAPI_BASE_URL}/watchers"
+CAPI_WATCHER_LOGIN_URL = f"{CAPI_BASE_URL}/watchers/login"
+CAPI_ENROLL_URL = f"{CAPI_BASE_URL}/watchers/enroll"
+CAPI_SIGNALS_URL = f"{CAPI_BASE_URL}/signals"
+CAPI_DECISIONS_URL = f"{CAPI_BASE_URL}/decisions/stream"
+CAPI_METRICS_URL = f"{CAPI_BASE_URL}/metrics"
+
+
+def machine_token_is_valid(machine: MachineModel) -> bool:
+    logging.debug(f"checking if token is valid for machine {machine.machine_id}")
+    try:
+        payload = jwt.decode(machine.token, options={"verify_signature": False})
+    except jwt.exceptions.DecodeError:
+        logging.debug(
+            f"could not decode token {machine.token} for machine {machine.machine_id}"
+        )
+        return False
+    current_time = time.time()
+    is_not_expired = current_time < payload["exp"]
+    logging.debug(
+        f"token for machine {machine.machine_id} is not expired = {is_not_expired}"
+    )
+    return is_not_expired
+
+
+class CAPIClient:
+    def __init__(
+        self, storage: StorageInterface, scenarios: List[str], max_retries: int = 3
+    ):
+        self.storage = storage
+        self.http_client = httpx.Client()
+        self.http_client.headers.update({"User-Agent": f"capi-py-sdk/{__version__}"})
+        self.scenarios = ",".join(sorted(scenarios))
+        self.max_retries = max_retries
+
+    def add_signals(self, signals: List[SignalModel]):
+        for signal in signals:
+            self.storage.update_or_create_signal(signal)
+
+    def send_signals(self, prune_after_send: bool = False):
+        unsent_signals_by_machineid = self._group_signals_by_machine_id(
+            filter(lambda signal: not signal.sent, self.storage.get_all_signals())
+        )
+        self._send_signals_by_machine_id(unsent_signals_by_machineid, prune_after_send)
+
+    def _group_signals_by_machine_id(
+        self, signals: Iterable[SignalModel]
+    ) -> Dict[str, List[SignalModel]]:
+        signals_by_machineid: Dict[str, List[SignalModel]] = defaultdict(list)
+        for signal in signals:
+            signals_by_machineid[signal.machine_id].append(signal)
+        return signals_by_machineid
+
+    def _send_signals_by_machine_id(
+        self,
+        signals_by_machineid: Dict[str, List[SignalModel]],
+        prune_after_send: bool = False,
+    ):
+        machines_to_process_attempts: List[MachineModel] = []
+        for machine_id, signals in signals_by_machineid.items():
+            machine_to_process = self._make_machine(
```

Comment:
[warn] move this in the  "while machines_to_process_attempts:" loop (CF other comment on the function itself)